### DISCUSSION TITLE
FreeCAD dependencies and how to manage them

### DISCUSSION THREAD

**Discussion by vejmarie** [Homebrew/discussions#342] [Everyday usage]
Hi,

I am maintaining FreeCAD homebrew formula. I am facing a challenge with boost. We extensively rely on it and the bottle that we generate are dependant of it. Last built was using 1.74.0 and we upgraded the core to 1.75 which broke FreeCAD on many end users Mac. Is their a way to be warned upfront as to syncing a new build of the bottles when such dependencies are upgraded ? That is a big pain to us. Could we postpone boost 1.74 to a formula form like boost@1.74 ? I tried that but it didn't exist.

vejmarie

**Comment by gromgit** [comment]
As far as I know, Homebrew expects casks to be self-contained, so you'd have to find a way to vendor the Boost libraries in your own app.

In any case, Homebrew is by nature an evergreen project, so what you're currently doing was bound to fail sooner or later.

**Comment by vejmarie** [comment]
I am not updating the Cask which is self contained with 0.18.4 release of FreeCAD, but the Formula which does contain the preview version of 0.19 and is getting some good success. Not afraid by failure, was wondering if their could be a way to implement callbacks when another formula from which I depend is updated, or if we could have fine grain management of formula version. I bet I am not the only one to face this challenge.

**Reply by carlocab** [reply]
>implement callbacks when another formula from which I depend is updated

This would be nice to know, and I'd be interested in this too. It might be possible to write a script that detects changes and run it in a cron-type job.

>implement callbacks when another formula from which I depend is updated

You can do this by maintaining your dependency in a tap. For example,

    brew extract --version=1.74.0 boost <tap>

and then replacing `depends_on "boost"` with `depends_on "boost@1.74.0"`. You can, of course, edit the formula to drop the trailing `.0`. You may also want to rename it to `freecad-boost` or something similar if the exact version isn't really what you care about.

**Reply by gromgit** [reply]
If I understand correctly, your use case is "install prebuild FreeCAD binaries from a DMG, all of which have linkages to Homebrew Boost libraries under `/usr/local`, instead of `/Applications/FreeCAD.app/Contents/Resources`". As far as I know, that's not a supported use case, so none of your requested features are likely to happen in an official release, though folks like @carlocab might be interested enough to hack something up.

**Reply by carlocab** [reply]
I actually suspect that @vejmarie is talking about this `freecad` formula: https://github.com/FreeCAD/homebrew-freecad/blob/master/Formula/freecad.rb

As for my interest, at this point, I think it's purely academic. I would like to know _how_ to do it. Whether I'd actually do it depends on how involved it would be. 😅

**Reply by gromgit** [reply]
> I actually suspect that @vejmarie is talking about this `freecad` formula: https://github.com/FreeCAD/homebrew-freecad/blob/master/Formula/freecad.rb

Ah, that one's new to me. For that, _definitely_ vendor your desired version with `brew extract` as @carlocab demonstrated, though you should probably fully-qualify the new dependency, i.e. `depends_on "freecad/freecad/boost@1.74.0"`.

You can also auto-build bottles to GitHub Releases with a recently-released Homebrew feature: https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/ . After all, Boost isn't a fast build for anyone. :fearful:

**Reply by vejmarie** [reply]
> If I understand correctly, your use case is "install prebuild FreeCAD binaries from a DMG, all of which have linkages to Homebrew Boost libraries under `/usr/local`, instead of `/Applications/FreeCAD.app/Contents/Resources`". As far as I know, that's not a supported use case, so none of your requested features are likely to happen in an official release, though folks like @carlocab might be interested enough to hack something up.

nop i am using  https://github.com/FreeCAD/homebrew-freecad/blob/master/Formula/freecad.rb which deploy in /usr/local

**Reply by carlocab** [reply]
>Boost isn't a fast build for anyone

Boost is actually pretty painless compared to Rust. Though that might've been because I was trying to compile it on a potato...

**Reply by vejmarie** [reply]
Ok I subscribed to the watch releases from homebrew core. Hopefully my mailbox is not going to be flooded too much.

**Reply by gromgit** [reply]
> Ok I subscribed to the watch releases from homebrew core. Hopefully my mailbox is not going to be flooded too much.

Quite the opposite: You'll get *nothing*, since GitHub's idea of a release is what software projects like yours associate with versions (i.e. 0.1, 0.2, etc.) Homebrew core taps are basically continuous streams of PRs and commits, with no release tags.

As I mentioned earlier, your best move is probably to _vendor_ Boost 1.74.0 via `brew extract`, and have your formula depend on that. Vendoring dependencies is a pretty common technique in modern software development.

**Reply by vejmarie** [reply]
I could to that, but I have so many dependencies ... (including Qt, which the formula failed to rebuild on my Mac) that is in the end it could end up to a massive "fork"

**Reply by gromgit** [reply]
Since your formula has Homebrew-managed dependencies, you'll have to either:
1. drop whatever you're doing at unexpected moments to fix issues against just-updated dependencies, or
2. vendor your dependencies and worry about upgrading when your core product is stable...or just ship with whatever versions are already working (and don't have horrible security issues)

I've dealt with developers using either track, and the latter ones usually make better progress.

That said, you can also choose to vendor only when a dependency upgrade breaks, rather than vendor everything at once. The choice is yours.

**Comment by SMillerDev** [comment]
> Is their a way to be warned upfront as to syncing a new build of the bottles when such dependencies are upgraded ?

Follow the upstream release channels. Homebrew is always going to be updated fairly quickly after an official release.

**Reply by vejmarie** [reply]
Loved to. Any idea on how to do it ? (sorry for the very basic question)

**Reply by carlocab** [reply]
Depends on upstream. If they have a GitHub repo, it's usually enough to watch the releases on their repo. If not, you may need to subscribe to their mailing list.
